[torch.package/TorchScript] logic to enable sharing of tensors on load

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57888 [PyTorch Edge] bytecode version bump to v5 and enable share constant table
* #58182 [PyTorch Edge] share tensors in mobile with new api
* #57678 [torch.package/TorchScript] flag to gate allowance of TS serializaiton in torch.package
* #54894 [torch.Package/TorchScript] TS into torch.Package test cases
* **#57573 [torch.package/TorchScript] logic to enable sharing of tensors on load**
* #54893 [torch.Package/TorchScript] torch.Package python logic to save TorchScript
* #56299 [torch.Package/TorchScript] ScriptModuleSerializer add unified format
* #55958 [torch.Package/TorchScript] refactor ScriptModuleSerializer Exporter
* #54891 [torch.Package/TorchScript] TS serialization importer to handle unified format

Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

Commit:
[torch.package/TorchScript] logic to enable sharing of tensors on load

[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"




[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"




[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Lilyjjo commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1044,6 +1044,26 @@ void initJITBindings(PyObject* module) {
+          "add_tensor_impl",
+          [](StorageContextTracker& self,
+             const std::string& name,
+             const at::Tensor& tensor) {
```
I'm not sure why, but I could only get `at::Tensor` and not `c10::Storage` to pass between Python and Cpp. The rest of the deserialization code looks as if it had the same issue. When looking at the uses of get_storage_from_record (the fucntion right above this one which is used for deserialization), people seem to use the tensor just as a vehicle for the storage also:
```
# In torch/serialization.py
 storage = zip_file.get_storage_from_record(name, size, dtype).storage()
```

suo replied:
I think this is because c10::Storage is bound directly to Python as opposed to using pybind, so the magic "just return it and we'll figure out how to turn it into the right python type" doesn't work. That's fine, it's probably not worth untangling rn.

Lilyjjo commented on torch/csrc/jit/serialization/import.h:
```diff
@@ -16,6 +17,35 @@ class ReadAdapterInterface;
+      std::string,
+      c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>>
+      storage_map_;
+};
```
I chose to use a vector of `c10::intrusive_ptr<c10::TensorImpl, c10::UndefinedTensorImpl>>` instead of a vector of `c10::Storages` because the `c10::TensorImpl` holds the `c10::Storage` in a ref-counted way. I'm still trying to verify that this way of tracking the storages doesn't lead to memory leaks. The comment at the top of the `TensorImpl` class also made it seem like the correct choice: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L331

suo replied:
I think c10::Storage is fine. It's a wrapper class that holds onto an underlying ref-counted StorageImpl. Let's use Storage here if we can to better capture intent.

Lilyjjo commented on torch/jit/_script.py:
```diff
@@ -341,9 +341,14 @@ def unpackage_script_module(importer: PackageImporter, script_module_id: str) ->
     Called by ``torch.package.PackageImporter``'s Pickler's ``persistent_load`` function.
     Performs work of loading and returning a ScriptModule from a ``torch.package`` archive.
     """
+    assert(isinstance(importer.zip_reader, torch._C.PyTorchFileReader))
```
TODO: I'm not sure how to handle the issue where TS objects can only be loaded when the `zip_reader` for a `PackageImporter` is a archive instead of a plain directory. This issue arrises because the TS deserialization code expects a `PyTorchFileReader`

suo replied:
what does this mean? I thought `zip_reader` is always a PyTorchFileReader or a MockZipReader.

Lilyjjo replied:
Yeah it's confusing. The MockZipReader isn't a PyTorchFileReader reader and The TorchScript deserialization code expects an actual PyTorchFileReader

Lilyjjo replied:
I added a better error message. I think we can fix this in a follow up PR

Lilyjjo commented on torch/package/package_importer.py:
```diff
@@ -170,14 +173,21 @@ def load_pickle(self, package: str, resource: str, map_location=None) -> Any:
+                tensor = self.zip_reader.get_storage_from_record(
+                    ".data/" + name, size, dtype
+                )
+                if isinstance(self.zip_reader, torch._C.PyTorchFileReader):
```
The `MockZipReader`'s returned 'storages' are of type 'torch.package._mock_zipreader._HasStorage' instead of an actual tensor , which is incompatible with the `StorageContextTracker''s requirement of needing a tensor passed between cpp and python. Due to the issue of TS also not being compatible with the `MockZipReader`, I don't think this will be a real issue

suo replied:
maybe add a comment describing what this is actually checking (for the case where `zip_reader` is a `MockZipReader`.

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

suo commented on torch/csrc/jit/serialization/import.h:
```diff
@@ -16,6 +17,35 @@ class ReadAdapterInterface;
 namespace jit {
 
+// used in torch.package deserialization
+class TORCH_API StorageContextTracker {
```
nit: StorageContextTracker seems redundant, why not just StorageContext

suo commented on torch/csrc/jit/serialization/unpickler.h:
```diff
@@ -48,7 +50,9 @@ class TORCH_API Unpickler {
-      bool use_storage_device = false)
+      bool use_storage_device = false,
+      std::shared_ptr<StorageContextTracker> storage_tracker =
+          std::shared_ptr<StorageContextTracker>(nullptr))
```
nit: I think this can just be `= nullptr`?

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

suo commented on torch/package/package_importer.py:
```diff
@@ -211,15 +221,17 @@ def persistent_load(saved_id):
-        def set_map_location():
-            # to let TorchScript's reduce_packge read specified map location
+        def set_deserialization_context():
+            # to let reduce_packge access deserializaiton context
```
reduce_package

suo reviewed (approved):
minor nits but overall lg!

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

cccclai commented on torch/csrc/jit/serialization/import.h:
```diff
@@ -16,6 +17,29 @@ class ReadAdapterInterface;
 namespace jit {
 
+// used in torch.package deserialization
+class TORCH_API StorageContext {
```
I doubt if we really need `StorageContext` here. Given that `root_key` is updated, the prefix can be added to the `root_key` to indicate the prefix. For the previous model, the `root_key` is the tensor index. After this format change, the `root_key` will be 
```
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '140671917317488.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
```
What I suggest is, make it 
```
```
    (torch._utils._rebuild_tensor_v2(pers.obj(('storage',
          torch.DoubleStorage,
          '{archive_name}/140671917317488.storage',
          'cpu',
          8),),
       0,
       (2, 4),
       (4, 1),
       False,
       collections.OrderedDict()),
```
Then we don't need to add different *_prefix. To handle old models, the default search path can be it's own archive. Anyway, it's just my suggestion to make it clean

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28226975).

Commit:
Update on "[torch.package/TorchScript] logic to enable sharing of tensors on load"


Changes in this PR enable the sharing of tensors between ScriptModules and eager modules when loading from a torch.package. See test case 'test_load_shared_tensors' and 'test_load_shared_tensors_repackaged' in #54894 to see it in action. Python IDs are not preserved, but the underlying TensorImpl are.

Differential Revision: [D28226975](https://our.internmc.facebook.com/intern/diff/D28226975)

[ghstack-poisoned]
